lib/concurrent-ruby/concurrent/atomic/constant_thread_local_var.rb

```diff
@@ -0,0 +1,45 @@
+require 'concurrent/atomic/abstract_thread_local_var'
+
+module Concurrent
+  # Has the same interface as {ThreadLocalVar} but the value can never be changed.
+  # The value is always the default provided to the constructor.
+  class ConstantThreadLocalVar < AbstractThreadLocalVar
+
+    # @note the value is always equal to default value
+    # @!macro thread_local_var_method_get
+    def value
+      default
+    end
+
+    # @!macro thread_local_var_method_set
+    # @raise ArgumentError if the value is not equal to the default
+    def value=(value)
+      if value != default
+        raise ArgumentError, "Constant thread local vars may not be altered"
+      end
+    end
+
+    # @!macro thread_local_var_method_bind
+    # @raise ArgumentError if the value is not equal to the default
+    def bind(value)
+      self.value = value
+      if block_given?
+        yield
+      end
+    end
+
+    protected
+
+    def allocate_storage
+      # nothing to do
+    end
+
+    def default
+      if @default_block
+        @default_block.call
```

Comment:
This evaluates the `@default_block` on each value read, so it may return different values each time value is called. I think that if the `default_block` is provided the ConstantThreadLocalVar should still store the computed value per thread (since it might differ). So the implementation could inherit from ThreadLocalVar, disallow using `value=` , and it would be good if it optimises `@default` case (the value is equal for all threads).